### ISSUE TITLE
Security Proposal: add "--storage-opt size=" to "docker service create/update" option list area/swarm

### ISSUE THREAD

**Issue by ghostplant** [docker/docker#28619]
This limit can prevent from running out of disk space by any one of applications inside the container using "dd if=/dev/zero of=/attack bs=1M", since "docker run" already supports this options.

**Comment by thaJeztah** [comment]
I linked this to https://github.com/docker/docker/issues/25303, which is where we keep track of options that are present on `docker run` but not (yet) supported for services

**Comment by barrachri** [comment]
Any update about this?
You can literally crash services without this option :)

**Comment by ghostplant** [comment]
@barrachri Still waiting for better solution. Currently I am using a fork version to protect this issue (--container-label com.docker.resource.storage.opt.size=2G).

**Comment by barrachri** [comment]
Hi @ghostplant, so you are passing `--container-label com.docker.resource.storage.opt.size=2G` to `docker create service` for now?

**Comment by ghostplant** [comment]
@barrachri Yes. But it only works with this patch(https://github.com/moby/moby/pull/31181/files) added.

**Comment by ghostplant** [comment]
@barrachri Actually, `storage.opt.size` is just one of them which can crash a service run by swarm. `pids`, `ulimit.opts`, `anonymous local volume` .. are also able to crash a swarm service.

**Comment by cirocosta** [comment]
Hey @ghostplant , and what has been your approach to these others you mentioned? Also in the fork?

**Comment by IvanBoyko** [comment]
It would be really nice if this was implemented.
It's currently the only thing that stops us from moving from standalone Windows containers to docker stack.
Really hoping you clever guys can find time for this :)

**Comment by ghostplant** [comment]
@IvanBoyko you can build docker project of your own to support these options~

**Comment by IvanBoyko** [comment]
You know I'd rather help professional to do it :)
Be it money I paid for Docker exam, or my readiness to work as beta-tester for Docker.

**Comment by andyfisher100** [comment]
Has any further work taken place on this? We are using swarm to provision windows containers that work as build machines but after about 4 builds the machines run out of disk space even though the host has 300GB free :(

**Comment by mhemrg** [comment]
Any progress on this?

**Comment by remorses** [comment]
I really need this feature, we already have Cpu and memory constraints but IMO disk size constraint is the most important

**Comment by combin** [comment]
Any news with it?

**Comment by markperri** [comment]
Has this been implemented or do I finally need to learn kubernetes?

**Comment by ghostplant** [comment]
@markperri You can build docker from source to patch this storage-size-limit feature.

**Comment by markperri** [comment]
@ghostplant Thanks, I will give it a try.  Will this work on the current version of docker?  Do I need to install it on both manager and worker nodes?
